### PR TITLE
feat: add solana gas fees endpoint

### LINKED ISSUE #1042: Solana - `/api/v1/gas/estimate`
### Overview

Placeholder - leverage Solana JSON-RPC endpoints to estimate gas for a given Tx. Note this is for computing units calculation, *not* for priority fee, which is to be its own endpoint.

### References and additional details

https://solana.com/docs/rpc/http/simulatetransaction
https://solana.com/docs/rpc/http/getfeeformessage

### Acceptance Criteria

- Simulate endpoints are happy
- We have slow, average and fast heuristics

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

3

### PR DESCRIPTION
**Pull request by NeOMakinG**
closes #1042 

## Description
Simply add the gas fee endpoint supporting the parameters `message`

as `getFeeForMessage` supports a `Message` param, it feels simpler to support the base64 of the message so we don't even have to build the full tx to estimate fees but only compose instructions and send them through this endpoint

We could also send the full base64 tx and decode the message part but why would we build the full TX to estimate gas fees while we can do it earlier? We can discuss about it if needed

We don't have a notion of slow, medium, fast gas fees, getFeeForMessage returns the best fees for the TX to be executed as fast as possible regarding the history of priority fees of the last TXs

## Screenshots
Small test using a basic send TX:
image

### CODE REVIEW

File: node/coinstacks/solana/api/src/models.ts [hunk 1]
```diff
@@ -51,3 +51,10 @@ export interface PriorityFees {
   // average confirmation speed estimation
   fast: number
 }
+
+/**
+ * Returns the gas fees estimation
+ */
+export interface GasFeesBody {
```
Comment (kaladinlight) [inline node/coinstacks/solana/api/src/models.ts:58]:
```suggestion
export interface EstimateFeesBody {
```

File: node/coinstacks/solana/api/src/models.ts [hunk 2]
```diff
@@ -51,3 +51,10 @@ export interface PriorityFees {
   // average confirmation speed estimation
   fast: number
 }
+
+/**
+ * Returns the gas fees estimation
```
Comment (kaladinlight) [inline node/coinstacks/solana/api/src/models.ts:56]:
```suggestion
 * Contains the base64 encoded transaction message
```

File: node/coinstacks/solana/api/src/controller.ts [hunk 3]
```diff
@@ -126,4 +127,30 @@ export class Solana implements BaseAPI {
       throw handleError(err)
     }
   }
+
+  /**
+   * Get the current recommended gas fees to use in a transaction
```
Comment (kaladinlight) [inline node/coinstacks/solana/api/src/controller.ts:132]:
```suggestion
   * Get the estimated compute unit cost of a transaction
```

File: node/coinstacks/solana/api/src/controller.ts [hunk 4]
```diff
@@ -126,4 +127,30 @@ export class Solana implements BaseAPI {
...
+   *
+   * @returns {Promise<number>} current fees specified in lamports including base fee
```
Comment (kaladinlight) [inline node/coinstacks/solana/api/src/controller.ts:134]:
```suggestion
   * @returns {Promise<number>} estimated compute unit cost
```

File: node/coinstacks/solana/api/src/controller.ts [hunk 5]
```diff
@@ -126,4 +127,30 @@ export class Solana implements BaseAPI {
...
+   */
+  @Response<BadRequestError>(400, 'Bad Request')
+  @Response<ValidationError>(422, 'Validation Error')
+  @Response<InternalServerError>(500, 'Internal Server Error')
+  @Post('/fees/estimate')
+  async getEstimateFees(@Body() body: GasFeesBody): Promise<number> {
```
Comment (kaladinlight) [inline node/coinstacks/solana/api/src/controller.ts:140]:
```suggestion
  async estimateFees(@Body() body: EstimateFeesBody): Promise<number> {
```

File: node/coinstacks/solana/api/src/controller.ts [hunk 6]
```diff
@@ -126,4 +127,30 @@ export class Solana implements BaseAPI {
...
+    try {
+      const deserializedMessage = Message.from(Buffer.from(body.message, 'base64'))
```
Comment (kaladinlight) [inline node/coinstacks/solana/api/src/controller.ts:142]: use versioned message to support both legacy and v0 messages
```suggestion
      const deserializedMessage = VersionedMessage.deserialize(Buffer.from(body.message, 'base64'))
```

File: node/coinstacks/solana/api/src/controller.ts [hunk 7]
```diff
@@ -126,4 +127,30 @@ export class Solana implements BaseAPI {
...
+
+      const feeResult = await heliusSdk.connection.getFeeForMessage(deserializedMessage)
+      const gasPrice = feeResult.value
+
+      if (!gasPrice) {
+        throw new Error('Failed to get gas price')
+      }
+
+      return gasPrice
```
Comment (kaladinlight) [inline node/coinstacks/solana/api/src/controller.ts:151]: destructure and update error to not include gas price vernacular
```suggestion
      const { value } = await heliusSdk.connection.getFeeForMessage(deserializedMessage)

      if (!value) throw new Error('Failed to get estimated fee')

      return value
```

**Comment by kaladinlight** [comment]
Overall looks sounds, just updates to vernacular and suggestion to use `VersionedMessage`.

We can update in the future to add `rawTx` as a possible param to estimate (simulate) from the raw unsigned tx depending on the most ergonomic usage in web.

**Review by kaladinlight — APPROVED** [review]

File: node/coinstacks/solana/api/src/controller.ts (hunk 4 above)
Comment (NeOMakinG) [reply node/coinstacks/solana/api/src/controller.ts]: Changed in [d50be64](https://github.com/shapeshift/unchained/pull/1051/commits/d50be6415f5b5822023b295cfe71938ed521ffb5)

File: node/coinstacks/solana/api/src/controller.ts (hunk 3 above)
Comment (NeOMakinG) [reply node/coinstacks/solana/api/src/controller.ts]: Changed in [d50be64](https://github.com/shapeshift/unchained/pull/1051/commits/d50be6415f5b5822023b295cfe71938ed521ffb5)

File: node/coinstacks/solana/api/src/models.ts (hunk 1 above)
Comment (NeOMakinG) [reply node/coinstacks/solana/api/src/models.ts]: Fixed in [d50be64](https://github.com/shapeshift/unchained/pull/1051/commits/d50be6415f5b5822023b295cfe71938ed521ffb5)

**Pull request merged**